### PR TITLE
Respect nixpkgs commit in devbox add by using nix search

### PR DESCRIPTION
**Pull request by savil**
## Summary

Devbox was not respecting the `nixpkgs.commit` from `devbox.json` for some operations
like `devbox info` and `devbox add`.

This PR fixes this by changing the `PkgInfo` function to rely on the `nix search` 
command, which lets us query for the exact package when given a nixpkgs commit.

For now, we strictly fail if the exact match was not found.  In the future, we could
improve the error message to suggest some of the specific alternatives that `nix search`
can provide.

Alternatives include:
1. `nix-env --json --file https://github.com/nixos/nixpkgs/archive/af9e00071d0971eb292fd5abef334e66eda3cb69.tar.gz  -qa pulumi`, but this takes 13 seconds.
2. Using `nix-instantiate`, but this doesn't let us later suggest specific alternatives like `nix search` does.
```
nix-instantiate --eval --strict --json --expr 'let
  pkgs = import
    (fetchTarball {
      url = "https://github.com/nixos/nixpkgs/archive/af9e00071d0971eb292fd5abef334e66eda3cb69.tar.gz";
    })
    { };
in { inherit (pkgs.pulumi-bin) name pname version system; }'
```

NOTE:
The current implementation will re-download the nixpkgs, even if `devbox shell` has downloaded it once. I anticipate that moving to `nix develop` and flakes will get rid of this problem.

## How was it tested?

```
❯ devbox info pulumi
Package pulumi not found
```

and

```
❯ devbox info pulumi-bin
pulumi-3.37.2-x86_64-darwin

```

Also, verified that `devbox add` now works, including inside a `devbox shell`.
```
testdata/rust/rust-stable on  savil/nix-search is 📦 v0.1.0 via 🦀 v1.61.0 on ☁️  (us-east-2) on ☁️  savil.srivastava@jetpack.io
❯ devbox shell
Installing nix packages. This may take a while... done.
Starting a devbox shell...
project dir is /Users/savil/code/jetpack/devbox/testdata/rust/rust-stable
info: using existing install for 'stable-x86_64-apple-darwin'
info: default toolchain set to 'stable-x86_64-apple-darwin'

  stable-x86_64-apple-darwin unchanged - rustc 1.64.0 (a55dd71d5 2022-09-19)

(devbox)
testdata/rust/rust-stable on  savil/nix-search is 📦 v0.1.0 via 🦀 v1.64.0 on ☁️  (us-east-2) on ☁️  savil.srivastava@jetpack.io
❯ devbox add pulumi-bin
Installing nix packages. This may take a while... done.
pulumi-bin is now installed. Run `hash -r` to ensure your shell is updated.
(devbox)
testdata/rust/rust-stable on  savil/nix-search [!] is 📦 v0.1.0 via 🦀 v1.64.0 on ☁️  (us-east-2) on ☁️  savil.srivastava@jetpack.io
❯ which pulumi
/usr/local/bin/pulumi
(devbox)
testdata/rust/rust-stable on  savil/nix-search [!] is 📦 v0.1.0 via 🦀 v1.64.0 on ☁️  (us-east-2) on ☁️  savil.srivastava@jetpack.io
❯ hash -r
(devbox)
testdata/rust/rust-stable on  savil/nix-search [!] is 📦 v0.1.0 via 🦀 v1.64.0 on ☁️  (us-east-2) on ☁️  savil.srivastava@jetpack.io
❯ which pulumi
/Users/savil/code/jetpack/devbox/testdata/rust/rust-stable/.devbox/nix/profile/default/bin/pulumi
```

Since `pulumi` doesn't exist in the stable `nixpkgs` channel, this would previously fail

### CODE REVIEW

**Comment by savil** [comment]
Current dependencies on/for this PR:
* main
  * **PR #314** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/314" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jetpack-io/devbox/314?utm_source=stack-comment).

File: nix/nix.go [hunk 1]
```diff
@@ -67,14 +71,35 @@ func parseInfo(pkg string, data []byte) *Info {
...
-	for _, result := range results {
+	for nixpkgs, result := range results {
 		pkgInfo := &Info{
 			NixName: pkg,
 			Name:    result["pname"].(string),
 			Version: result["version"].(string),
-			System:  result["system"].(string),
 		}
+
+		reLegacyPackages := regexp.MustCompile(fmt.Sprintf("legacyPackages\\.(.*)\\.%s", pkg))
```
[the remark this answers is not in the document]
Comment (gcurtis) [reply nix/nix.go:81]: This seems weird, but when I search for a linux-only package such as `fuse3` I get:

```
> nix search nixpkgs#fuse3 --json
{"legacyPackages.aarch64-darwin.fuse3":{"pname":"fuse","version":"3.11.0","description":"Library that allows filesystems to be implemented in user space"}}
```

But I get an error if I try to install it, which is what I'd expect:

```
> nix-env -iA nixpkgs.fuse3
installing 'fuse-3.11.0'
error: Package ‘fuse-3.11.0’ in /nix/store/37nl1pnk837xd0033qw7jrq7i18q36vn-nixpkgs/nixpkgs/pkgs/os-specific/linux/fuse/common.nix:92 is not supported on ‘aarch64-darwin’, refusing to evaluate.
```

Does `nix search` not actually consider the system?

**Review by gcurtis — APPROVED** [review]

File: nix/nix.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (savil) [reply nix/nix.go:81]: Oh good catch, I think there's a flag I can use to filter for system

File: nix/nix.go [hunk 2]
```diff
@@ -43,11 +45,13 @@ func Exec(path string, command []string) error {
...
 }
 
-func PkgInfo(pkg string) (*Info, bool) {
+func PkgInfo(nixpkgsCommit, pkg string) (*Info, bool) {
 	buf := new(bytes.Buffer)
-	attr := fmt.Sprintf("nixpkgs.%s", pkg)
-	cmd := exec.Command("nix-env", "--json", "-qa", "-A", attr)
+	exactPackage := fmt.Sprintf("nixpkgs/%s#%s", nixpkgsCommit, pkg)
+	cmd := exec.Command("nix", "search", "--json", exactPackage)
+	cmd.Args = appendExperimentalFeatures(cmd.Args, "nix-command", "flakes")
```
Comment (loreto) [inline nix/nix.go:52]: When `nix search` needs to download the nixpkgs repository, it shows some download progress. Are we showing it too? I'm wondering if we need to do `cmd.Stderr = os.Stderr`

File: nix/nix.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (savil) [reply nix/nix.go:81]: I was wrong about the flag....

[the remark this answers is not in the document]
Comment (savil) [reply nix/nix.go:81]: hmm wouldn't we want to check that the package exists for the user's current system?

**Pull request merged**